feat: revamp store & upload specs (#114)
web3-storage/specs: w3-store.md

# W3 Storage Protocol

> ![status:wip](https://img.shields.io/badge/status-wip-orange.svg?style=flat-square)

Reviewer: SHould we state that this is in deprecation stage?

Author: I don't think we should deprecate until we have alternative

> ![status:wip](https://img.shields.io/badge/status-wip-orange.svg?style=flat-square)

Author: ```suggestion
![reliable](https://img.shields.io/badge/status-reliable-green.svg?style=flat-square)
```

## Authors

- [Irakli Gozalishvili](https://github.com/gozala)
- [Yusef Napora](https://github.com/yusefnapora), [DAG House](https://dag.house)

## Abstract

In the W3 protocol user owned (name)space represent a data storage primitive that can be managed using W3 storage protocol defined here. Storage protocol allows space owners to manage state across compatible storage provider services using defined set of [UCAN] capabilities. Use of [UCAN] authorization system enables space access to be shared by delegating corresponding capabilities to desired audience.

## Language

The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in [RFC2119](https://datatracker.ietf.org/doc/html/rfc2119).

# Introduction

The base storage layer of the space is [Content Address]ed [Content ARchive][CAR] files. In practice this means that user wishing to store files or directories of files needs produce an [IPLD] [DAG] in [UnixFS] format and then encoded into one or multiple [Content ARchive]s that can be stored using W3 storage protocol.

> > Large DAGs get "sharded" across multiple [Content ARchive]s and stored individually to meet size limits of the storage provider.

Reviewer: ```suggestion
> Large DAGs get "sharded" across multiple [Content ARchive]s and stored individually to meet potential size limits of the storage provider.
```

well, by our experience with old.web3.storage, I think we also concluded that more shards can be better for paralelization and more efficient retries on error. In addition, this also enables building a pause and resume upload pipeline which is great use case. I would suggest expanding this as not only a potential limitation of the storage provider

##### Store Add Origin

> Status: Deprecated 🛑
>
> Field was intended for establish causal order, but proved impractical.

Capability invocation MAY set _optional_ `nb.origin` field to a causally related [CAR] like a previous shard of the content DAG.

##### Store Add Example

```js
{
  "v": "0.9.1",
  "iss": "did:key:z6Mkk...xALi",
  "aud": "did:web:web3.storage",
  "att": [
    {
      can: "store/add",
      with: "did:key:zAl..1ce",
      nb: {
        link: { "/": "bag...7ldq" },
        size: 42_600_000
      }
    }
  ],
  "prf": [{ "/": "bafy...prf1" }],
  "exp": 1685602800,
  "s": {
    "/": {
      "bytes": "7aEDQJbJqmyMxTxcK05XQKWfvxG+Tv+LWCJeE18RSMnciCZ/RQ21U75LA0uFSvIjdqnF5RaauZTE8mh2ZYMBBejdJQ4"
    }
  }
}
```

#### Store Add Receipt

Provider MUST issue signed receipt containing `StoreAddResult`.

##### Store Add Result

###### Store Add Result IPLD Schema

```ipldsch
type StoreAddResult union {
  StoreAddSuccess   "ok"
  StoreAddFailure   "error"
} representation keyed

type StoreAddSuccess union {
  StoreAddDone      "done"
  StoreAddPending   "upload"
} representation inline {
  discriminantKey "status"
}

type StoreAddDone struct {
  with    SpaceDID

>   link    &ContentArchive

Reviewer: we miss allocated property (this is used for billing to get to know how much to bill to the space). see:
- https://github.com/web3-storage/w3up/blob/main/packages/upload-api/src/store/add.js#L67
- https://github.com/web3-storage/w3up/pull/1213

}

type StoreAddPending struct {
  with      SpaceDID

>   link      &ContentArchive

Reviewer: we miss allocated property (this is used for billing to get to know how much to bill to the space). see:
- https://github.com/web3-storage/w3up/blob/main/packages/upload-api/src/store/add.js#L67
- https://github.com/web3-storage/w3up/pull/1213

  with      SpaceDID
  nb        StoreGet
}

type StoreGet struct {
  link            &ContentArchive
}
```

##### Store Get Content

Capability invocation MUST specify [Content archive Identifier].

##### Store Get Example

```js
{
  "v": "0.9.1",
  "iss": "did:key:z6Mkk...xALi",
  "aud": "did:web:web3.storage",
  "att": [
    {
      can: "store/get",
      with: "did:key:zAl..1ce",
      nb: {
        link: { "/": "bag...7ldq" },
      }
    }
  ],
  "prf": [{ "/": "bafy...prf1" }],
  "exp": 1685602800,
  "s": {
    "/": {
      "bytes": "7aEDQJbJqmyMxTxcK05XQKWfvxG+Tv+LWCJeE18RSMnciCZ/RQ21U75LA0uFSvIjdqnF5RaauZTE8mh2ZYMBBejdJQ4"
    }
  }
}
```

#### Store Get Receipt

Capability provider MUST issue signed receipt containing `StoreGetResult`.

##### Store Get Result

###### Store Get Result IPLD Schema

```ipldsch
type StoreGetResult struct {
  StoreGetSuccess   "ok"
  StoreGetFailure   "error"
} representation keyed
```

###### Store Get Success

Capability provider MUST issue `StoreGetSuccess` result for every content archive that has been added to the space.

```ipldsch
type StoreGetSuccess {
  link            &ContentArchive
  size            Int

>   origin optional &ContentArchive

Reviewer: should we comment that it is deprecated?

>   origin optional &ContentArchive

Author: ```suggestion
  # deprecated
  origin optional &ContentArchive
```

#### Store Remove Capability

##### Store Remove IPLD Schema

```ipldsch
type StoreRemoveCapability struct {
  with      SpaceDID
  nb        StoreRemove
}

type StoreRemove struct {
  link            &ContentArchive
}
```

##### Store Remove Content

Capability invocation MUST specify desired [Content archive Identifier].

##### Store Remove Example

```js
{
  "v": "0.9.1",
  "iss": "did:key:z6Mkk...xALi",
  "aud": "did:web:web3.storage",
  "att": [
    {
      can: "store/remove",
      with: "did:key:zAl..1ce",
      nb: {
        link: "bag...7ldq",
      }
    }
  ],
  "prf": [{ "/": "bafy...prf1" }],
  "exp": 1685602800,
  "s": {
    "/": {
      "bytes": "7aEDQJbJqmyMxTxcK05XQKWfvxG+Tv+LWCJeE18RSMnciCZ/RQ21U75LA0uFSvIjdqnF5RaauZTE8mh2ZYMBBejdJQ4"
    }
  }
}
```

#### Store Remove Receipt

Capability provider MUST issue signed receipt containing `StoreRemoveResult`.

##### Store Remove Result

###### Store Remove Result IPLD Schema

```ipldsch
type StoreRemoveResult struct {
  StoreRemoveSuccess   "ok"
  StoreRemoveFailure   "error"
} representation keyed
```

###### Store Remove Success

> Capability provider MUST issue `StoreRemoveSuccess` after it has removed the archive from the space.

Reviewer: ```suggestion
Capability provider MUST issue `StoreRemoveSuccess` after it unlinked the archive from the space.
```
I think using unlined would be better, so that it does not get strictly understood as content won't be available anymore on the network

```ipldsch
type StoreListFailure struct {
  message       string
}
```

## Upload Capabilities

Capabilities under `upload/*` namespace can be used to manage list of top level content entries. While not required, it is generally assumed that user content like file will be turned into [UnixFS] DAG packed and stored in space as one or more [content archive][CAR]s. The root of the DAG is then added to the upload list.

### Upload Capabilities IPLD Schema

```ipldsch
type UploadCapability union {
  UploadAddCapability      "upload/add"
  UploadGetCapability      "upload/get"
  UploadListCapability     "upload/list"
  UploadRemoveCapability   "upload/remove"
} representation inline {
  discriminantKey "can"
}
```

### Upload Add

Authorized agent MAY invoke `upload/add` capability on the [space] subject (`with` field) to request that provider (`aud` field) include content identified by `nb.root` in the list of content entries for the space.

It is expected that CARs containing content are stored in the space using [`store/add`] capability. Provider MAY enforce this invariant by failing invocation or choose to succeed invocation but fail to serve the content when requested.

> > ⚠️ Behavior of calling `upload/add` with a same `root` and different `shards` is not specified by the protocol. w3up reference implementation allows such invocations and updates `shards` to union of all shards across invocations.

Reviewer: maybe worth adding a note that there may exist limit of shards at implementation level?

      nb: {
        root: { "/": "bafy...ro0t" },
      }
    }
  ],
  "prf": [{ "/": "bafy...prf1" }],
  "exp": 1685602800,
  "s": {
    "/": {
      "bytes": "7aEDQJbJqmyMxTxcK05XQKWfvxG+Tv+LWCJeE18RSMnciCZ/RQ21U75LA0uFSvIjdqnF5RaauZTE8mh2ZYMBBejdJQ4"
    }
  }
}
```

#### Upload Get Receipt

Capability provider MUST issue signed receipt containing `UploadGetResult`.

##### Upload Get Result

###### Upload Get Result IPLD Schema

```ipldsch
type UploadGetResult struct {
  UploadGetSuccess   "ok"
  UploadGetFailure   "error"
} representation keyed
```

###### Upload Get Success

Capability provider MUST issue `UploadGetSuccess` result for the upload entry that has been added to the space.

```ipldsch
type UploadGetSuccess {
  link            &any
  shards          [&ContentArchive]
}
```

###### Upload Get Failure

Capability provider MUST issue `UploadGetFailure` result if content entry has not been added or was since removed from space at the time of the invocation.

```ipldsch
type UploadGetFailure {
  message   string
}
```

### Upload Remove

Authorized agent MAY invoke `upload/remove` capability to remove upload entry from the list in the specified space (`with` field).

> > ⚠️ Please note that removing upload entry SHOULD NOT remove [content archive][CAR]s containing contain.

Reviewer: should this be a MUST instead?

Author: ```suggestion
> ⚠️ Please note that removing upload entry MUST NOT remove [content archive][CAR]s containing contain.
```
